### DISCUSSION TITLE
[PROPOSAL] Split the core functionalities into NPM packages

### DISCUSSION THREAD

**Discussion by raisiqueira** [eruptionjs/core#53] [Ideas]
To keep all Eruption packages on the same line, I propose to create NPM packages using the [`core`](https://github.com/eruptionjs/core) configurations in the existent projects (`core`, `react-lib` and `website`), also in future packages, like the `opinionated` EruptionJS. Furthermore, with that, we can plug the Eruption config into any existent project 🤩.

## Proposed packages:

- `@eruptionjs/prettier-rules`: The rules used in the `core` project for React applications.
- `@eruptionjs/stylelint-rules`: The CSS/SCSS rules for those projects that want to use CSS (like TailwindCSS) instead of a CSS-In-JS solution.
- `@eruptionjs/tsconfig-rules`: Our strict and well configured TypeScript config.
- `@eruptionjs/eslint-rules`: The base ESLint rules.
  - `@eruptionjs/eslint-rules/react`: The ESLint rules for React projects.

I see it as an advantage, since we can also handle these packages on existent projects using our CLI (https://github.com/eruptionjs/create-eruption) through a `update-project` command.

We can start a test with the Prettier package, since I see it as an easy way to start, not too much configuration.

Thoughts, @luigieai @devmozao?

![eruptionjs-packages-mindmap](https://github.com/eruptionjs/core/assets/2914170/c594a941-2a4a-471b-992d-6ec60a44353d)

**Comment by luigieai** [comment]
Seems cool, in a long term vision, what we can plan to take advantage of the core modularization?

**Reply by devmozao** [reply]
Seems a solid evolution and also would help on pushing updates to the projects that implements eruption. Let's talk about it on how to. Do you have something in mind, @raisiqueira ?

**Reply by raisiqueira** [reply]
Yes, I have. I'm planning to start with a package with our Prettier rules, or even with our ESLint rules, since we have an [issue](https://github.com/eruptionjs/core/issues/44) to improving the ESLint rules from the `core` repository.

The idea is:

- Start a monorepo using PNPM and [NX](https://nx.dev) with a local registry ([Verdeccio](https://verdaccio.org/)) to test it locally before publishing on NPM. Note: I already use this stack on a [personal package](https://github.com/raisiqueira/headless-stepper-monorepo). 
- After that, setup GitHub action to auto-publish these packages to our organization on NPM.
